swim/self_evict.go

```diff
@@ -0,0 +1,244 @@
+package swim
+
+import (
+	"errors"
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/ringpop-go/util"
+
+	"github.com/uber-common/bark"
+)
+
+var (
+	// ErrDuplicateHook is returned when a hook with the given name has already
+	// been registered.
+	ErrDuplicateHook = errors.New("expected hook name to be unique")
+
+	// ErrEvictionInProgress is returned when ringpop is already in the process
+	// of evicting
+	ErrEvictionInProgress = errors.New("ringpop is already executing an eviction")
+)
+
+// SelfEvict defines the functions that interact with the self eviction of nodes
+// from the membership prior to shutting down
+type SelfEvict interface {
+	RegisterSelfEvictHook(hooks SelfEvictHook) error
+	SelfEvict() error
+}
+
+// SelfEvictHook is an interface describing a module that can be registered to
+// the self eviction hooks
+type SelfEvictHook interface {
+	// Name returns the name under which the eviction should be regitered
+	Name() string
```

Comment:
Fyi; names have been removed now to simplify the interface. We prevent double registration the same way we do for EventListeners (by equality).